[SPARK-14709][ML] spark.ml API for linear SVM

hhbyyh opened the pull request:
## What changes were proposed in this pull request?

jira: https://issues.apache.org/jira/browse/SPARK-14709 

Provide API for SVM algorithm for DataFrames. As discussed in jira, the initial implementation uses OWL-QN with Hinge loss function.
The API should mimic existing spark.ml.classification APIs.
Currently only Binary Classification is supported. Multinomial support can be added in this or following release.
## How was this patch tested?

new unit tests and simple manual test

hhbyyh committed:
prototype

hhbyyh committed:
Merge remote-tracking branch 'upstream/master' into mlsvm

hhbyyh committed:
Merge remote-tracking branch 'upstream/master' into mlsvm

hhbyyh committed:
svm prototype

SparkQA commented:
**[Test build #65816 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65816/consoleFull)** for PR 15211 at commit [`f8ddc3b`](https://github.com/apache/spark/commit/f8ddc3bd03ae81e9a4fe2462e83aa4288418cdb5).

hhbyyh commented:
@yanboliang this is a quick prototype for the ML SVM. I'll make another pass tomorrow to refine the code and add more unit tests. It mimics many behavior as LR, yet I'm not sure if we need to achieve similar complexity as LR . 
For reviewers, appreciate if you can take a glance just in case there's something out of expectation. We can work on details after I remove the WIP tag.

SparkQA commented:
**[Test build #65816 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65816/consoleFull)** for PR 15211 at commit [`f8ddc3b`](https://github.com/apache/spark/commit/f8ddc3bd03ae81e9a4fe2462e83aa4288418cdb5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SVM @Since(\"2.1.0\") (`
  * `  class SVMModelWriter(instance: SVMModel)`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65816/
Test PASSed.

YY-OnCall committed:
Merge remote-tracking branch 'upstream/master' into mlsvm

hhbyyh committed:
Merge remote-tracking branch 'upstream/master' into mlsvm

hhbyyh committed:
param cleanup

SparkQA commented:
**[Test build #65898 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65898/consoleFull)** for PR 15211 at commit [`73b8011`](https://github.com/apache/spark/commit/73b8011ff9edef4fc96fdf68951b990f2fa340f6).

SparkQA commented:
**[Test build #65898 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65898/consoleFull)** for PR 15211 at commit [`73b8011`](https://github.com/apache/spark/commit/73b8011ff9edef4fc96fdf68951b990f2fa340f6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65898/
Test PASSed.

yanboliang commented on mllib/src/main/scala/org/apache/spark/ml/classification/SVM.scala:
```diff
@@ -0,0 +1,527 @@
+ */
+@Since("2.1.0")
+@Experimental
+class SVM @Since("2.1.0") (
```
What about ```SVMClassifier```? we can also train regression model with SVM.

hhbyyh replied:
I changed it to LinearSVC, so we can have other SVM Classifier in the future.

yanboliang commented on mllib/src/main/scala/org/apache/spark/ml/classification/SVM.scala:
```diff
@@ -0,0 +1,527 @@
+@Experimental
+class SVM @Since("2.1.0") (
+    @Since("2.1.0") override val uid: String)
+  extends Predictor[Vector, SVM, SVMModel]
```
Under the framework ```Classifier``` or ```ProbabilisticClassifier```?

hhbyyh replied:
Thanks I changed it to Classifier. AFAIK, SVM raw result may not be used to indicate the probability.

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/classification/SVM.scala:
```diff
@@ -0,0 +1,527 @@
+  with SVMParams with DefaultParamsWritable {
+
+  @Since("2.1.0")
+  def this() = this(Identifiable.randomUID("svm"))
```
if we rename this to `SVMClassifier`, the uid should be `svc`

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/classification/SVM.scala:
```diff
@@ -0,0 +1,527 @@
+      }
+
+    val instr = Instrumentation.create(this, instances)
+    instr.logParams(regParam, standardization, threshold, maxIter, tol, fitIntercept)
```
To keep in line with other algos, `labelCol`, `weightCol`, `featuresCol` should be added here

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/classification/SVM.scala:
```diff
@@ -0,0 +1,527 @@
+        n
+      case None => histogram.length
+    }
+
```
`require(numClasses == 2, "...")` ?

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/classification/SVM.scala:
```diff
@@ -0,0 +1,527 @@
+      val localCoefficientsArray = coefficientsArray
+      val localGradientSumArray = gradientSumArray
+
+      numClasses match {
```
what about moving checking of `numClasses` outside of `add()`?

hhbyyh commented:
Thanks for suggestions @yanboliang @zhengruifeng. I'm thinking to rename to class to LinearSVMClassifier. Just as other linearSVM implementation, this will only be a binary classifier and multi-classification will be supported via one-vs-rest. I'll also add Penalty parameter `C`. For `dual`, I'm thinking maybe we should just set it to true.

YY-OnCall committed:
Merge remote-tracking branch 'upstream/master' into mlsvm

YY-OnCall committed:
rename to linearsvc

YY-OnCall committed:
Merge remote-tracking branch 'upstream/master' into mlsvm

YY-OnCall committed:
merge

YY-OnCall committed:
style update

SparkQA commented:
**[Test build #70155 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70155/testReport)** for PR 15211 at commit [`4902517`](https://github.com/apache/spark/commit/4902517f82de547a1984c369764c52ddb96dcc9b).

SparkQA commented:
**[Test build #70155 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70155/testReport)** for PR 15211 at commit [`4902517`](https://github.com/apache/spark/commit/4902517f82de547a1984c369764c52ddb96dcc9b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70155/
Test PASSed.

hhbyyh commented:
Sent an update to address some comments.

hhbyyh commented on mllib/src/main/scala/org/apache/spark/ml/classification/SVM.scala:
```diff
@@ -0,0 +1,527 @@
+  override protected def train(dataset: Dataset[_]): SVMModel = {
+    val w = if (!isDefined(weightCol) || $(weightCol).isEmpty) lit(1.0) else col($(weightCol))
+    val instances: RDD[Instance] =
+      dataset.select(col($(labelCol)).cast(DoubleType), w, col($(featuresCol))).rdd.map {
```
Yes, thanks.

YY-OnCall committed:
Merge remote-tracking branch 'upstream/master' into mlsvm

YY-OnCall committed:
add more ut and small refine

SparkQA commented:
**[Test build #70232 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70232/testReport)** for PR 15211 at commit [`c8a7553`](https://github.com/apache/spark/commit/c8a75532f049e16186d63ae8eb22f4ea51eb1cde).

SparkQA commented:
**[Test build #70232 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70232/testReport)** for PR 15211 at commit [`c8a7553`](https://github.com/apache/spark/commit/c8a75532f049e16186d63ae8eb22f4ea51eb1cde).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70232/
Test PASSed.

hhbyyh commented:
Remove WIP.

YY-OnCall committed:
rename weight to coefficient

SparkQA commented:
**[Test build #70279 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70279/testReport)** for PR 15211 at commit [`6bab7c9`](https://github.com/apache/spark/commit/6bab7c9f5fc39653458364ce424645c1ab6de79f).

SparkQA commented:
**[Test build #70279 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70279/testReport)** for PR 15211 at commit [`6bab7c9`](https://github.com/apache/spark/commit/6bab7c9f5fc39653458364ce424645c1ab6de79f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70279/
Test PASSed.

jkbradley commented:
Thanks for the PR!  I'll take a look

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala:
```diff
@@ -0,0 +1,525 @@
+  extends ClassificationModel[Vector, LinearSVCModel]
+  with LinearSVCParams with MLWritable {
+
+  override val numClasses = 2
```
Also add ```override val numFeatures: Int```

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala:
```diff
@@ -0,0 +1,525 @@
+  def setThreshold(value: Double): this.type = set(threshold, value)
+
+  /**
+   * Predict label for the given features.
```
Remove doc.  There's no need to copy the doc for overridden methods unless the doc is specialized for this class.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala:
```diff
@@ -0,0 +1,525 @@
+   * This internal method is used to implement [[transform()]] and output [[predictionCol]].
+   */
+  override protected def predict(features: Vector): Double = {
+    val margin = BLAS.dot(features, coefficients) + intercept
```
Abstract the margin computation here and in predictRaw into a private method.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala:
```diff
@@ -0,0 +1,525 @@
+  }
+
+  /**
+   * Returns a [[org.apache.spark.ml.util.MLWriter]] instance for this ML instance.
```
remove copied doc

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala:
```diff
@@ -0,0 +1,525 @@
+    override protected def saveImpl(path: String): Unit = {
+      // Save metadata and Params
+      DefaultParamsWriter.saveMetadata(instance, path, sc)
+      // Save model data: numClasses, numFeatures, intercept, coefficients
```
Remove doc (which is incorrect); there's no need for it.

jkbradley reviewed (commented):
I've reviewed the public API, but not the implementation yet.  Feel free to go ahead and update it though.

One high-level comment: I'd prefer to call this SVMClassifier so that we do not introduce a duplicate API in the future for multiclass SVMs.  The only reason to split the APIs would be if we think we will not be able to extend this one, but I believe it will be a simple API change (adding a Param for specifying the kernel).

hhbyyh commented:
Thanks a lot for the review. @jkbradley 

About the class name. AFAIK, typically "linear SVM" and "general SVM" use different algorithms for implementations. Just like the difference between [LIBSVM](https://www.csie.ntu.edu.tw/~cjlin/libsvm/) and [LIBLINEAR](https://www.csie.ntu.edu.tw/~cjlin/liblinear/), also as [LinearSVC](http://scikit-learn.org/stable/modules/generated/sklearn.svm.LinearSVC.html) and [SVC](http://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html) in sklearn. LinearSVC only support SVM classifier with linear kernel and usually has a better scalability, while SVC support different kernels and does not really scale well in LIBSVM or sklearn. In a way, LinearSVC is a special acceleration for general SVM. It has independent public API because it uses fundamentally different techniques.

I imagine some day in the future, we perhaps need to provide a new implementation for SVM with kernel. Let me know if you still prefer to have a unified interface for SVM.

jkbradley commented:
You have a good point about setting user expectations about speed and scalability.  I don't think that the average user needs to understand the underlying implementation, but performance expectations are important.  Let's go with what you have (LinearSVC).

Side note: If we wanted to match libsvm/liblinear, then we would not add a new class but would just add hinge loss support to GeneralizedLinearRegression.  I hesitate to do that since GLMs technically require natural exponential families and adding hinge loss would make it harder to explain algorithm behavior (such as which evaluation stats are available).

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala:
```diff
@@ -0,0 +1,178 @@
+package org.apache.spark.ml.classification
+
+import breeze.linalg.{DenseVector => BDV}
+import scala.util.Random
```
organize imports

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala:
```diff
@@ -0,0 +1,178 @@
+  }
+
+  test("params") {
+    ParamsSuite.checkParams(new LogisticRegression)
```
fix: not logreg

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala:
```diff
@@ -0,0 +1,178 @@
+    assert(!lsvc.isDefined(lsvc.weightCol))
+    assert(lsvc.getFitIntercept)
+    assert(lsvc.getStandardization)
+    val model = lsvc.fit(smallBinaryDataset)
```
set maxIter to 2 for speed

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala:
```diff
@@ -0,0 +1,178 @@
+    assert(model.getRawPredictionCol === "rawPrediction")
+    assert(model.intercept !== 0.0)
+    assert(model.hasParent)
+
```
check numFeatures too

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala:
```diff
@@ -0,0 +1,178 @@
+  test("read/write: SVM") {
+    def checkModelData(model: LinearSVCModel, model2: LinearSVCModel): Unit = {
+      assert(model.intercept === model2.intercept)
+      assert(model.coefficients.toArray === model2.coefficients.toArray)
```
No need for "toArray"

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala:
```diff
@@ -0,0 +1,178 @@
+    val A = 0.01
+    val B = -1.5
+    val C = 1.0
+    val binaryDataset = {
```
Why not use smallBinaryDataset?

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala:
```diff
@@ -0,0 +1,178 @@
+object LinearSVCSuite {
+
+  val allParamSettings: Map[String, Any] = Map(
+    "regParam" -> 0.01,
```
set predictionCol, rawPredictionCol, aggregationDepth to non-defaults too

hhbyyh committed:
Merge remote-tracking branch 'upstream/master' into mlsvm

hhbyyh commented on mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala:
```diff
@@ -0,0 +1,525 @@
+  override val numClasses = 2
+
+  @Since("2.2.0")
+  def setThreshold(value: Double): this.type = set(threshold, value)
```
Not sure I get it... how do we use weight in the prediction

jkbradley replied:
True...not useful yet.  It will be in the future if we add an evaluate() method to return a summary for a new dataset, but it's better not to have it now.

hhbyyh committed:
add more setters, refine API and comments

SparkQA commented:
**[Test build #70539 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70539/testReport)** for PR 15211 at commit [`21ecbf0`](https://github.com/apache/spark/commit/21ecbf08ed03f9b69f4cbec7380e547f146acec7).

SparkQA commented:
**[Test build #70539 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70539/testReport)** for PR 15211 at commit [`21ecbf0`](https://github.com/apache/spark/commit/21ecbf08ed03f9b69f4cbec7380e547f146acec7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class LinearSVC @Since(\"2.2.0\") (`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70539/
Test PASSed.

hhbyyh commented:
I've sent a new update addressing most of the comments. The only exception is about `SetWeightCol` in `LinearSVCModel`. cc @jkbradley.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala:
```diff
@@ -0,0 +1,525 @@
+
+/**
+ * :: Experimental ::
+ * Linear SVM Classifier with Hinge Loss and OWLQN optimizer
```
This too please!

jkbradley replied:
Oh, I see you did the change for LinearSVCModel.  Could you please do the same for LinearSVC?

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala:
```diff
@@ -0,0 +1,558 @@
+
+  /**
+   * Set the convergence tolerance of iterations.
+   * Smaller value will lead to higher accuracy at the cost of more iterations.
```
"value" -> "values"

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala:
```diff
@@ -0,0 +1,558 @@
+  @Since("2.2.0")
+  override val numFeatures: Int = coefficients.size
+
+    @Since("2.2.0")
```
fix indent

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala:
```diff
@@ -0,0 +1,558 @@
+  def setThreshold(value: Double): this.type = set(threshold, value)
+
+  @Since("2.2.0")
+  def setWeightCol(value: Double): this.type = set(threshold, value)
```
You were right; we don't need this yet.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala:
```diff
@@ -0,0 +1,558 @@
+  @Since("2.2.0")
+  def setWeightCol(value: Double): this.type = set(threshold, value)
+
+  private val margin: Vector => Double = (features) => {
```
It's apparently more efficient to write a "def" than a "val" though I know vals are used in some other models.

hhbyyh replied:
Please confirm. I thought using "val" will be more efficient, as "def" will create a new instance of the function each time it's invoked.

jkbradley replied:
Yeah, maybe what I'd been told was incorrect: I just looked on SO and it sounds like you are correct.  val sounds good to me.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala:
```diff
@@ -0,0 +1,558 @@
+  setDefault(tol -> 1E-6)
+
+  /**
+   * whether to standardize the training features before fitting the model.
```
"whether" -> "Whether"

jkbradley commented:
Btw, @hhbyyh @yanboliang do either of you know of resources proving convergence rates for OWL-QN for the hinge loss?  The original OWL-QN paper's analysis does not handle hinge loss.

jkbradley commented:
Btw, I'm making a review pass now for the internals.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala:
```diff
@@ -0,0 +1,558 @@
+      val rawCoefficients = state.x.toArray.clone()
+      val coefficientArray = Array.tabulate(numFeatures) { i =>
+        // flatIndex will loop though rawCoefficients, and skip the intercept terms.
+        val flatIndex = if ($(fitIntercept)) i + i / numFeatures else i
```
This is overly complicated logic copied from logreg.  Simplify for this binary case.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala:
```diff
@@ -0,0 +1,558 @@
+         Note that the intercept in scaled space and original space is the same;
+         as a result, no scaling is needed.
+       */
+      val rawCoefficients = state.x.toArray.clone()
```
Why do we need to clone the array?  It isn't modified below.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala:
```diff
@@ -0,0 +1,558 @@
+      val optimizer = new BreezeOWLQN[Int, BDV[Double]]($(maxIter), 10, regParamL1Fun, $(tol))
+      val initialCoefficientsWithIntercept = Vectors.zeros(numFeaturesPlusIntercept)
+      if ($(fitIntercept)) {
+        initialCoefficientsWithIntercept.toArray(numFeatures) = math.log(
```
This initialization is calculated for the logistic loss, not the hinge loss.  Have you looked in literature for a good initialization for hinge loss?

jkbradley replied:
Thinking about this, I think it should be initialized to +1 if there are more positive than negative examples and to -1 otherwise.  (Write out the loss function and split it into these 2 sums to see this.)  @yanboliang @dbtsai Does that sound reasonable?

Could you please check out other libraries to see what they do?  Thanks!

jkbradley reviewed (commented):
OK, done with a review pass through the internals.  I still need to check the tests though.

Noting some follow-ups for future JIRAs:
* user guide
* Python API
* R API
* model summary
* corner cases not yet handled:
  * constant label
    * including with and without fitIntercept
    * including with constant-value non-0 column

Related to my question about the convergence of OWL-QN on hinge loss, do we have any reason to believe that OWL-QN will be better than L-BFGS for hinge loss?  OWL-QN is analyzed w.r.t. non-differentiable regularizers, but not losses, AFAIK.

hhbyyh committed:
Merge remote-tracking branch 'upstream/master' into mlsvm

hhbyyh commented on mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala:
```diff
@@ -0,0 +1,558 @@
+  private var weightSum = 0.0
+  private var lossSum = 0.0
+
+  private val coefficientsArray = coefficients match {
```
Is it possible that we use Sparse Vector for coefficients in the future?

jkbradley replied:
That's what I was thinking.

hhbyyh commented on mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala:
```diff
@@ -0,0 +1,558 @@
+   */
+  def add(instance: Instance): this.type = {
+    instance match { case Instance(label, weight, features) =>
+      require(weight >= 0.0, s"instance weight, $weight has to be >= 0.0")
```
The validation should has been checked before with MultivariateOnlineSummarizer.

jkbradley replied:
You're right, and now that I look, the weight will also have been validated by MultivariateOnlineSummarizer

hhbyyh committed:
Merge remote-tracking branch 'upstream/master' into mlsvm

hhbyyh commented on mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala:
```diff
@@ -0,0 +1,558 @@
+ */
+private class LinearSVCCostFun(
+    instances: RDD[Instance],
+    numClasses: Int,
```
I left it for possible multi class support. But sure we can add it back if necessary.

hhbyyh committed:
code simplification and add necessary check

SparkQA commented:
**[Test build #70684 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70684/testReport)** for PR 15211 at commit [`07725b9`](https://github.com/apache/spark/commit/07725b99998088d9a980a91dc757540ba96a65ca).

hhbyyh commented:
Thanks Joseph. I've addressed most comments except a few exceptions. I'll follow up to see the default intercept in linearSVM and sklearn.

SparkQA commented:
**[Test build #70684 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70684/testReport)** for PR 15211 at commit [`07725b9`](https://github.com/apache/spark/commit/07725b99998088d9a980a91dc757540ba96a65ca).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70684/
Test PASSed.

jkbradley commented:
Thanks!  The updates look good.  I'll check out the unit tests now.

Thanks for looking into the default intercept.  Also, let me know if you find literature about convergence analyses of OWL-QN or L-BFGS on hinge loss.

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala:
```diff
@@ -0,0 +1,166 @@
+import org.apache.spark.ml.util.{DefaultReadWriteTest, MLTestingUtils}
+import org.apache.spark.ml.util.TestingUtils._
+import org.apache.spark.mllib.util.MLlibTestSparkContext
+import org.apache.spark.sql.{Dataset, Row}
```
no need to import Row

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala:
```diff
@@ -0,0 +1,554 @@
+      case None => histogram.length
+    }
+    require(numClasses == 2, s"LinearSVC only supports binary classification." +
+      s" $numClasses classes detected in $labelCol")
```
"$labelCol" -> "column $labelCol"

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala:
```diff
@@ -0,0 +1,166 @@
+
+    // Generate noisy input of the form Y = signum(x.dot(weights) + intercept + noise)
+  def generateSVMInput(
+    intercept: Double,
```
fix indent (+2)

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala:
```diff
@@ -0,0 +1,166 @@
+        Array.fill[Double](weights.length)(rnd.nextDouble() * 2.0 - 1.0))
+    val y = x.map { xi =>
+      val yD = new BDV(xi).dot(weightsMat) + intercept + 0.01 * rnd.nextGaussian()
+      if (yD < 0) 0.0 else 1.0
```
nit: Write as ```if (yD > 0) 1.0 else 0.0``` to match LinearSVCModel.predict when yD = 0

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala:
```diff
@@ -0,0 +1,166 @@
+    assert(lsvc.getRawPredictionCol === "rawPrediction")
+    assert(!lsvc.isDefined(lsvc.weightCol))
+    assert(lsvc.getFitIntercept)
+    assert(lsvc.getStandardization)
```
As long as we're doing this, how about checking maxIter, tol, aggregationDepth?

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala:
```diff
@@ -0,0 +1,166 @@
+    MLTestingUtils.checkCopy(model)
+  }
+
+  test("linear svc doesn't fit intercept when fitIntercept is off") {
```
Can this test use maxIter = 2?
Also, to be self-contained, this test should check that fitting with an intercept creates a non-0 intercept, though I know that is also checked elsewhere.

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala:
```diff
@@ -0,0 +1,166 @@
+    smallValidationDataset = generateSVMInput(A, Array[Double](B, C), nPoints, 17).toDF()
+  }
+
+  test("Linear SVC binary classification") {
```
This and the next test are decent sanity checks, but they aren't enough to give much confidence in the implementation.  Would you mind adding a unit test which compares with R's svm from the e1071 package, like we've done with some other algorithms?

(I know we don't have that in spark.mllib, but it'd be great to add here.)

jkbradley reviewed (commented):
Done checking unit tests!

YY-OnCall committed:
Merge remote-tracking branch 'upstream/master' into mlsvm

YY-OnCall committed:
Merge branch 'mlsvm' of https://github.com/hhbyyh/spark into mlsvm

YY-OnCall committed:
Merge remote-tracking branch 'upstream/master' into mlsvm

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala:
```diff
@@ -0,0 +1,554 @@
+private[classification] trait LinearSVCParams extends ClassifierParams with HasRegParam
+  with HasMaxIter with HasFitIntercept with HasTol with HasStandardization with HasWeightCol
+  with HasThreshold with HasAggregationDepth {
+
```
nit, the braces here can be removed

YY-OnCall committed:
Merge remote-tracking branch 'upstream/master' into mlsvm

YY-OnCall committed:
fix intercept update

SparkQA commented:
**[Test build #70913 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70913/testReport)** for PR 15211 at commit [`8e1d674`](https://github.com/apache/spark/commit/8e1d67481cd1e9ad9f935f10e1a95b06d5379a32).

hhbyyh commented:
Thanks @jkbradley  and @zhengruifeng. Just sent an update to fix the intercept update. 

I further evaluated the implementation and found setting intercept to 0 for initialization works better. That makes some sense as we can just regard intercept as a part of coefficients and give the same initial value as other coefficients.

SparkQA commented:
**[Test build #70913 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70913/testReport)** for PR 15211 at commit [`8e1d674`](https://github.com/apache/spark/commit/8e1d67481cd1e9ad9f935f10e1a95b06d5379a32).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70913/
Test PASSed.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala:
```diff
@@ -495,17 +495,16 @@ private class LinearSVCAggregator(
         features.foreachActive { (index, value) =>
           if (localFeaturesStd(index) != 0.0 && value != 0.0) {
-            localGradientSumArray(index) += value * gradientScale / localFeaturesStd(index)
+            localGradientSumArray(index) -= value * labelScaled * weight / localFeaturesStd(index)
```
Is it not better to compute this multiplication outside of the loop?

YY-OnCall committed:
Merge remote-tracking branch 'upstream/master' into mlsvm

YY-OnCall committed:
update ut

SparkQA commented:
**[Test build #71105 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71105/testReport)** for PR 15211 at commit [`05fbd02`](https://github.com/apache/spark/commit/05fbd0243db9466eb773c1ea1594fdbc9efe3c84).

SparkQA commented:
**[Test build #71105 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71105/testReport)** for PR 15211 at commit [`05fbd02`](https://github.com/apache/spark/commit/05fbd0243db9466eb773c1ea1594fdbc9efe3c84).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71105/
Test PASSed.

hhbyyh commented:
Sent an update to include a R unit test. Yet I met a problem that there's a constant scaling difference between LinearSVC and R 1071 (which essentially is LibSVM). It's possible that it's caused by some parameter setting. Post it anyway to see if there's any suggestions.

Sorry @zhengruifeng, I'll address your comment in the next update.

YY-OnCall committed:
Merge remote-tracking branch 'upstream/master' into mlsvm

jkbradley reviewed (commented):
For comparing with R, I'm wondering if the main issue is that it's hard to calculate the appropriate C given a regParam setting.  Would it be easier to use this R package instead?  https://cran.r-project.org/web/packages/svmpath/

Also, the test with sample weights takes 40 seconds.  Does it still pass if you increase the 'tol' Param to make the test faster?

YY-OnCall committed:
Merge remote-tracking branch 'upstream/master' into mlsvm

YY-OnCall committed:
Merge remote-tracking branch 'upstream/master' into mlsvm

YY-OnCall committed:
update ut

YY-OnCall committed:
remove trait bracket

SparkQA commented:
**[Test build #71618 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71618/testReport)** for PR 15211 at commit [`36f585c`](https://github.com/apache/spark/commit/36f585c14bdf82d7c35c22917842f06f937fcc8f).

hhbyyh commented:
Thanks @jkbradley.

I reviewed the gradient and loss function. The corresponding regularization lambda should satisfy lambda * c * N (data size) = 2. 
c equals to 10 by default in R and sklearn. thus we should set reg = 0.00002  (2 / 10000 / 10).

I corrected the unit test against R 1071 and added another unit test against sklearn.

SparkQA commented:
**[Test build #71619 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71619/testReport)** for PR 15211 at commit [`72719fc`](https://github.com/apache/spark/commit/72719fc62b7c1a55d1947defc8027dfe68627d42).

YY-OnCall committed:
update unit test path and remove require

SparkQA commented:
**[Test build #71620 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71620/testReport)** for PR 15211 at commit [`2e99a0f`](https://github.com/apache/spark/commit/2e99a0fe701c5b3415fdcad2a27e0ea4c0435f36).

SparkQA commented:
**[Test build #71618 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71618/testReport)** for PR 15211 at commit [`36f585c`](https://github.com/apache/spark/commit/36f585c14bdf82d7c35c22917842f06f937fcc8f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71618/
Test PASSed.

SparkQA commented:
**[Test build #71619 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71619/testReport)** for PR 15211 at commit [`72719fc`](https://github.com/apache/spark/commit/72719fc62b7c1a55d1947defc8027dfe68627d42).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71619/
Test PASSed.

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala:
```diff
@@ -0,0 +1,251 @@
+    assert(model1.coefficients ~== coefficientsR relTol 1E-2)
+  }
+
+  test("linearSVC comparison with scikit-learn") {
```
Let's combine these to avoid retraining since this is the same as the R test.  (And training takes 35 sec)

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala:
```diff
@@ -0,0 +1,251 @@
+      features <- as.matrix(data.frame(data$V2, data$V3, data$V4, data$V5))
+      svm_model <- svm(features, label, type='C', kernel='linear', cost=10, scale=F, tolerance=1e-4)
+      summary(svm_model)
+      w <- -t(svm_model$coefs) %*% svm_model$SV
```
Remove "-" to make values match lines coefficientsR, interceptR below

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala:
```diff
@@ -0,0 +1,251 @@
+      > w
+             data.V2   data.V3   data.V4   data.V5
+      [1,] -7.310475 -14.89742 -22.21019 -29.83495
+      > -svm_model$rho
```
same here

hhbyyh replied:
"-" here may be necessary as b = -model$rho #Offset

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala:
```diff
@@ -0,0 +1,251 @@
+
+  test("linearSVC comparison with scikit-learn") {
+    val trainer1 = new LinearSVC()
+      .setRegParam(0.00002)
```
Add a comment that this matches C above.

jkbradley reviewed (commented):
@hhbyyh This looks great.  Combining the tests to reduce test time is the only remaining issue, I believe.

SparkQA commented:
**[Test build #71620 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71620/testReport)** for PR 15211 at commit [`2e99a0f`](https://github.com/apache/spark/commit/2e99a0fe701c5b3415fdcad2a27e0ea4c0435f36).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71620/
Test PASSed.

hhbyyh commented:
I see, will work on combining the tests now. Also I'm thinking if we should consider using `c` (cost) to replace `RegParam` in `LinearSVC' to be more friendly for SVM users. Yet the change may be confusing for Spark users. I'm neutral on this.

YY-OnCall committed:
combine ut

SparkQA commented:
**[Test build #71623 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71623/testReport)** for PR 15211 at commit [`d62c107`](https://github.com/apache/spark/commit/d62c107400209bd6f31e0da76e254e76a47f6447).

YY-OnCall committed:
add comment for regParam

SparkQA commented:
**[Test build #71625 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71625/testReport)** for PR 15211 at commit [`9b71a3a`](https://github.com/apache/spark/commit/9b71a3ad4f64411bcab032628eb97736bc71adcf).

SparkQA commented:
**[Test build #71623 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71623/testReport)** for PR 15211 at commit [`d62c107`](https://github.com/apache/spark/commit/d62c107400209bd6f31e0da76e254e76a47f6447).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71623/
Test PASSed.

SparkQA commented:
**[Test build #71625 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71625/testReport)** for PR 15211 at commit [`9b71a3a`](https://github.com/apache/spark/commit/9b71a3ad4f64411bcab032628eb97736bc71adcf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71625/
Test PASSed.

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala:
```diff
@@ -0,0 +1,241 @@
+
+      > w
+             data.V2   data.V3   data.V4   data.V5
+      [1,] -7.310338 -14.89741 -22.21005 -29.83508
```
w and the intercept are still negative, which isn't what we wanted, right?

hhbyyh replied:
The updated weights are the correct ones and they are stable. I forgot how to generate the original weights... maybe use a different random seed for data generation.

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala:
```diff
@@ -0,0 +1,241 @@
+      [1] -7.440177
+
+     */
+    val coefficientsR = Vectors.dense(7.310475, 14.89742, 22.21019, 29.83495)
```
Why are these values changed?

hhbyyh replied:
The updated weights are the correct ones and they are stable. I forgot how to generate the original weights... maybe use a different random seed for data generation.

jkbradley reviewed (commented):
I'd like to keep regParam.  I think it's about as common in literature and practice as specifying the constraint C.

YY-OnCall committed:
Merge remote-tracking branch 'upstream/master' into mlsvm

YY-OnCall committed:
change R ut

SparkQA commented:
**[Test build #71742 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71742/testReport)** for PR 15211 at commit [`bbcb7cb`](https://github.com/apache/spark/commit/bbcb7cbcd17badced9936dfe0c8df24ae9059ceb).

SparkQA commented:
**[Test build #71742 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71742/testReport)** for PR 15211 at commit [`bbcb7cb`](https://github.com/apache/spark/commit/bbcb7cbcd17badced9936dfe0c8df24ae9059ceb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71742/
Test PASSed.

jkbradley commented:
LGTM
Thanks @hhbyyh and also @yanboliang and @zhengruifeng for helping with review!
Merging with master

One more step towards feature parity for the DataFrame-based API!

jkbradley commented:
I'll create follow-up JIRAs (linked from this PR's JIRA).  @hhbyyh Can I assign one or more to you?

hhbyyh commented:
Thanks @jkbradley for driving the review process. 
Also thanks @yanboliang and @zhengruifeng for the helpful comments.
Sure I'd like to keep working on the follow-up tasks.
